geopmdpy/geopmdpy/read.py

```diff
@@ -0,0 +1,96 @@
+core                        {topo.num_domain('core')}
+cpu                         {topo.num_domain('cpu')}
+memory                      {topo.num_domain('memory')}
+package_integrated_memory   {topo.num_domain('package_integrated_memory')}
+nic                         {topo.num_domain('nic')}
+package_integrated_nic      {topo.num_domain('package_integrated_nic')}
+gpu                         {topo.num_domain('gpu')}
+package_integrated_gpu      {topo.num_domain('package_integrated_gpu')}
+gpu_chip                    {topo.num_domain('gpu_chip')}""")
+
+def print_info(signal_name):
+    print(f'{signal_name}:\n    {pio.signal_description(signal_name)}')
+
+def print_info_all():
+    for signal_name in pio.signal_names():
+        print_info(signal_name)
+
+def print_signals():
+    print('\n'.join(pio.signal_names()))
+
+def run():
+    parser = ArgumentParser(description=__doc__)
+    parser_group = parser.add_mutually_exclusive_group()
+    parser_group.add_argument('-d', '--domain', action='store_true',
+                              help='print domains detected')
+    parser_group.add_argument('-i', '--info',
+                              help='print longer description of a signal')
+    parser_group.add_argument('-I', '--info-all', action='store_true',
+                              help='print longer description of all signals')
+    parser_group.add_argument('-c', '--cache', action='store_true',
+                              help='Create geopm topo cache if it does not exist')
+    parser_group.add_argument('-v', '--version', action='store_true',
+                              help='Print version and exit.')
+    positional_group = parser.add_argument_group()
+    positional_group.add_argument('SIGNAL_NAME', nargs='?',
+                                  help='Name of signal')
+    positional_group.add_argument('DOMAIN_TYPE',  nargs='?',
+                                  help='Name of the domain for which the signal should be read')
+    positional_group.add_argument('DOMAIN_INDEX', nargs='?', type=int,
+                                  help='Index of the domain, starting from 0')
+    args = parser.parse_args()
+    if args.domain:
+        print_domains()
+    elif args.info:
+        print_info(args.info)
+    elif args.info_all:
+        print_info_all()
+    elif args.cache:
+        topo.create_cache()
+    elif args.version:
+        print(__version_str__)
+    elif args.SIGNAL_NAME is not None and args.DOMAIN_TYPE is not None and args.DOMAIN_INDEX is not None:
+        signal = pio.read_signal(args.SIGNAL_NAME, args.DOMAIN_TYPE, args.DOMAIN_INDEX)
+        print(pio.format_signal(signal, pio.signal_info(args.SIGNAL_NAME)[1]))
+    elif args.SIGNAL_NAME is None and args.DOMAIN_TYPE is None and args.DOMAIN_INDEX is None:
+        print_signals()
+    else:
+        parser.print_help()
+        return -1
+    return 0
+
+def main():
+    err = 0
+    try:
+        err = run()
+    except RuntimeError as ee:
```

Comment:
What's special about RuntimeError? Why not catch all Exception types or none of them?

Reply:
This is copy pasta from all of our command line interfaces.  The original thinking (maybe not a great thought) was that we use RuntimeError for things that are "bad, but not unexpected if the runtime environment is not set up right", where as other errors are "bad, and unexpected so better emit the stack trace".  

Now, is this true of all our thrown exceptions in the code base, maybe not.